src/runtimes/rust/builder.js

```diff
@@ -0,0 +1,49 @@
+const { readFile } = require('fs')
+const { basename, join } = require('path')
+const { promisify } = require('util')
+
+const pReadFile = promisify(readFile)
+
+const tmp = require('tmp-promise')
+const toml = require('toml')
+
+const { lstat } = require('../../utils/fs')
+const { runCommand } = require('../../utils/shell')
+
+const { BUILD_TARGET, MANIFEST_NAME } = require('./constants')
+
+const build = async ({ srcDir }) => {
+  // We compile the binary to a temporary directory so that we don't pollute
+  // the user's functions directory.
+  const { path: targetDirectory } = await tmp.dir()
+  const functionName = basename(srcDir)
+
+  try {
+    await runCommand('cargo', ['build', '--target', BUILD_TARGET, '--release'], {
+      cwd: srcDir,
+      env: {
+        CARGO_TARGET_DIR: targetDirectory,
```

Comment:
you'll want to think about caching this to reduce build times in the future.
we do this for builds if the root of the repo is a Cargo project: https://github.com/netlify/build-image/blob/xenial/run-build-functions.sh#L701-L704

building a rust project from scratch is usually quite expensive, so levering the cache should be a big win for most people.

for our own projects we use sccache with an S3 bucket, but that seems a bit overkill since it doesn't integrate well with our regular caching strategy.

Reply:
I did think about that and I'm glad you brought it up.

I'm using a temporary directory here because zip-it-and-ship-it might be running on people's machines, in which case we don't want to be injecting files into their project directories.

But we could use a similar strategy to the one we use for builds. When running on buildbot, zip-it-and-ship-it could adopt a deterministic target directory (e.g. `.netlify/functions-rust-cache/{FUNCTION-NAME}`) and we'd add `.netlify/functions-rust-cache` to the list of directories that we cache in between builds.

And now that I think of it, this behaviour would also make sense in the local development environment, since we want functions to build as quickly as possible in response to changes. The key is that we use a directory that we know will be git-ignored (like `.netlify/`).

What do you think?